### DISCUSSION TITLE
Sjoin produces too many output geometries

### DISCUSSION THREAD

**Discussion by AlexanderJuestel** [geopandas/geopandas#2774] [Q&A] [answered]
Hello, 

I have a line that intersects multiple polygons (see the image below, the color of the line represents the length of each segment within a given polygon, I have used the overlay method here already). The aim is to calculate the length of the line section within each polygon and attribute the length of the line segment to the polygon to calculate a sort of "line density" (there will be many more lines per polygon in the end). When using `sjoin`, I would expect four geometries as output. However, I am getting 10. Some of the lines are counted for two polygons (the left and the right segments) and the middle ones are counted for three polygons. I have attached the sample data. I cannot wrap my head around it why they are counted multiple times. I also checked if the polygons overlap but could not find anything. 

Any hint is appreciated!
Cheers
Alex

Code to reproduce the issue:

```python

import geopandas as gpd

# Load files
line = gpd.read_file('line.shp')
mask_gdf = gpd.read_file('mask_gdf.shp')

# Intersect line with polygons
data_intersected = gpd.overlay(line, mask_gdf, how='intersection')
data_intersected['length'] = data_intersected.length

# Perform leftjoin
leftjoin_gdf = gpd.sjoin(left_df=data_intersected,
                             right_df=mask_gdf,
                             how="left")

```

I later want to use `gdf_grouped = (leftjoin_gdf.groupby('index_right')['length'].sum())` and `gdf_length = pd.concat([gdf_grouped, mask_gdf], axis=1)` to sum up the lengths in each polygon.

![image](https://user-images.githubusercontent.com/45469915/216083247-a6dde6e0-f45f-4055-9749-e26844c6c62f.png)
[data.zip](https://github.com/geopandas/geopandas/files/10558576/data.zip)

**Comment by AlexanderJuestel** [comment]
It seems similar to #2773

**Comment by brendan-ward** [comment, marked as answer]
@AlexanderJuestel this is because the overlay (intersection) produces 4 line segments from the original single segment, and then you join those 4 against the 4 boxes.

The sjoin uses `intersects` predicate by default, so your 2 outer segments each intersect 2 boxes, and your inner 2 segments each intersect 3 boxes => 10 combinations.  Each segment intersects the box that it falls within, as well as the box that it touches at its endpoint.

In this case, you could use the `within` predicate instead:

```
leftjoin_gdf = gpd.sjoin(left_df=data_intersected,right_df=mask_gdf,how="left", predicate='within')
```

Since you are aggregating to the boxes, you can change the order of operations a bit, and join first then calculate intersection and length of segment within each box:

```Python
joined = gpd.sjoin(left_df=mask_gdf, right_df=line).join(line.geometry.rename('line'), on='index_right')
# joined now has 'geometry' for boxes and 'line' for line geometries

joined['length'] = joined.geometry.intersection(joined.line).length

# to get total lengths per box, because joined is indexed on the same index as mask_gdf, so we use level=0
total_length_per_box = joined.groupby(level=0).length.sum()
```

**Reply by AlexanderJuestel** [reply]
Hey @brendan-ward, thanks for the clarification! Your proposed solution works: 

![image](https://user-images.githubusercontent.com/45469915/216117045-331864c8-2312-4567-bc82-e9e8118aad8b.png)

And also for the task that I was doing ;) 
![image](https://user-images.githubusercontent.com/45469915/216122423-95b96c18-6c00-46f3-8556-3bf51dfb3b7c.png)

**Reply by AlexanderJuestel** [reply]
It also works perfectly with a self-implemented hexagon grid (the maritime zones were not used here)

![image](https://user-images.githubusercontent.com/45469915/216130226-1b36719e-6fc3-45e5-a895-a5c2232424be.png)
